### ISSUE TITLE
Never fully hiding in Jest tests

### ISSUE THREAD

**Issue by skrivanos** [atomiks/tippyjs-react#252]
It seems like Tippy never actually hides and unmounts when using jest + testing-library. onHide is called, but the test fails on `waitForElementToBeRemoved` and `onHidden` is also never called.

```
import React from 'react';
import * as test from '@testing-library/react';
import Tippy from '@tippyjs/react';

it('should hide', async () => {
  const onHide = jest.fn();
  const onHidden = jest.fn();
  const { rerender, queryByText } = test.render(
    <Tippy content="foo" onHide={onHide} onHidden={onHidden} visible />,
  );
  expect(queryByText('foo')).toBeInTheDocument();
  rerender(<Tippy content="foo" onHide={onHide} onHidden={onHidden} visible={false} />);
  expect(onHide).toHaveBeenCalled();
  await test.waitForElementToBeRemoved(() => queryByText('foo'));
  expect(onHidden).toHaveBeenCalled();
});
```

```
  ● should hide

    Timed out in waitForElementToBeRemoved.

      12 |   rerender(<Tippy content="foo" onHide={onHide} onHidden={onHidden} visible={false} />);
      13 |   expect(onHide).toHaveBeenCalled();
    > 14 |   await test.waitForElementToBeRemoved(() => queryByText('foo'));
         |              ^
      15 |   expect(onHidden).toHaveBeenCalled();
      16 | });
      17 |

      at Object.waitForElementToBeRemoved (../node_modules/@testing-library/dom/dist/wait-for-element-to-be-removed.js:22:24)
      at Object.<anonymous> (__tests__/tippy.test.tsx:14:14)
```
Version: 4.1.0
TippyJS version: 6.2.6

**Comment by atomiks** [comment]
Is it due to the transitions maybe?

**Comment by skrivanos** [comment]
> Is it due to the transitions maybe?

Still happens with `animation={false}` though.

**Comment by ewilson-rx** [comment]
@skrivanos did you find a resolution for this? I'm currently running into seemingly the same issue

**Comment by skrivanos** [comment]
@ewilson-rx No, simply refraining from testing hiding of Tippy-based components at the moment...

**Comment by ewilson-rx** [comment]
@skrivanos dang. thanks anyway!

**Comment by cscheffauer** [comment]
Facing same issue unfortunately.

**Comment by cscheffauer** [comment]
@atomiks could someone of you guys take a look at this? would be great to have those hiding tests for Tippy-based components. Thanks :)

**Comment by KubaJastrz** [comment]
Hi @cscheffauer and everyone, 

I hope to find some time this weekend for this. Do you have some reproductions available? Does this issue occur while using the core tippy.js too? Or is it just the react wrapper?

**Comment by KubaJastrz** [comment]
So it turns out that the problem is in CSS transition that is applied when the tooltip is closing. If you know how to wait until they are done, then you can do that. But I'm not sure if this is possible with JSDOM

Alternatively you can set `animation={false}` or `duration={0}` if `process.env.NODE_ENV === 'test'`, so it's stripped away in production builds. (The env variable is provided by `jest`.)

**Comment by skrivanos** [comment]
@KubaJastrz Doesn't seem to be true for controlled mode... The following test case fails for me with the latest versions of react-testing-library, Tippy and Tippyjs-react:

```
import React from 'react';
import * as test from '@testing-library/react';
import Tippy from '@tippyjs/react';

it('should hide', async () => {
  const onHide = jest.fn();
  const onHidden = jest.fn();
  const { rerender, queryByText } = test.render(
    <Tippy
      content="foo"
      onHide={onHide}
      onHidden={onHidden}
      animation={false}
      delay={0}
      visible
    />,
  );
  expect(queryByText('foo')).toBeInTheDocument();
  rerender(
    <Tippy
      content="foo"
      onHide={onHide}
      onHidden={onHidden}
      animation={false}
      delay={0}
      visible={false}
    />,
  );
  expect(onHide).toHaveBeenCalled();
  await test.waitForElementToBeRemoved(() => queryByText('foo'));
  expect(onHidden).toHaveBeenCalled();
});
```

```
Timed out in waitForElementToBeRemoved.
```

**Comment by KubaJastrz** [comment]
There must be a version mismatch somewhere, as it seems like the same code can either pass or fail, based on the environment (system?).

I've created a [reproduction on codesandbox](https://codesandbox.io/s/tippy-jest-repro-6jikh) which fails in the browser (it uses `jest` directly), but passes locally (with `react-scripts test`).

I was suspecting the `jsdom` dependency to be the cause, but it seems like the code passes on both [`14.1.0`](https://github.com/KubaJastrz/tippy.js-react/tree/unmount-test) and [`16.6.0`](https://github.com/KubaJastrz/tippy.js-react/tree/jsdom-16) jsdom versions.

**Comment by cscheffauer** [comment]
I've experienced the following:
The hiding test fails when clicking on the children of the Tippy.
The hiding test works when the tippy is controlled via `instance.hide()` from a parent.

**Comment by cscheffauer** [comment]
@KubaJastrz just wanted to check in - is someone of you guys actively working on it? I can also take a closer look at it if u like :) cheers

**Comment by KubaJastrz** [comment]
Hey @cscheffauer, the further investigation needs to be done, I haven't worked on this since our last conversation.

I'm still suspecting something fishy going on with the jest setup, and not the tippy itself, but if you have time to create a reproduction and test a few things that would be greatly appreciated!

Please let me know if this is a blocking issue for you.

**Comment by cscheffauer** [comment]
It's still blocking us yes - we are not able to run the hiding tests at the moment. I will be able to take a look at this in 2-3 weeks probably. Keep you posted!

**Comment by atomiks** [comment]
I've tracked the bug down to a line in `tippy.js` core. `unmount()` never fully gets executed because of this line:

https://github.com/atomiks/tippyjs/blob/475b30829095bce8f79e44f3d77b558bb8f35cb0/src/createTippy.ts#L1100-L1102

Which means the `isMounted` state is not being updated correctly when the tippy mounts. It gets set to `true` here, which is what it needs to run the whole unmount operation:

https://github.com/atomiks/tippyjs/blob/475b30829095bce8f79e44f3d77b558bb8f35cb0/src/createTippy.ts#L1003

This is an async part, called when the popper instance gets created and `onFirstUpdate` is called for the very first time, but it's not being called by jest because it gets re-rendered with a `false` state before that happens. Popper does this via a `Promise` microtask. 

If you wait for the mounting process to have occurred fully asynchronously, it should work.

**Comment by cscheffauer** [comment]
@atomiks can u pls bump tippyjs-react to use the latest version of tippy? the fix u have made in [#1003](https://github.com/atomiks/tippyjs/pull/1003) doesn't show up in the react package. Thanks!

**Comment by atomiks** [comment]
@cscheffauer just remove then reinstall the package and it'll update the transitive `tippy.js` dependency (on both Yarn & npm)

**Comment by cscheffauer** [comment]
@atomiks Thanks - that updated the Tippy.js package. Still tough it didn't fix the test i've been working on unfortunately:
![image](https://user-images.githubusercontent.com/29776408/149295927-9f142164-d673-44af-ad5a-5bb4656b7e26.png)
The last assertion when hiding the Popover again isn't passing. All other assertions work fine till that point.
The component uses LazyTippy underneath, animation is set to false.

**Comment by atomiks** [comment]
Does it work with `fireEvent` instead of `userEvent`?

**Comment by cscheffauer** [comment]
same behaviour as with `userEvent` unfortunately

**Comment by atomiks** [comment]
Does the example I gave work? I'd try to narrow it down by removing abstractions over <Tippy /> to find the root cause. The "async" bug should be gone now though.

**Comment by cscheffauer** [comment]
i think the example isn't 100% the same scenario i'm testing. In the scenario the tippy will be hidden programmatically from the parent component (which I've been able to test successfully through exposing the instance object).
In my test I'm using RTL to click on a button and trigger the tippy from there. 

Something I just noticed - the second `userEvent.click()` in my example test causes this error message in the console:
`Error: Uncaught [Error: requestAnimationFrame is not supported in Node]`
